test_conformance/extensions/cl_khr_command_buffer/negative_command_buffer_finalize.cpp

```diff
@@ -0,0 +1,134 @@
+//
+// Copyright (c) 2024 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "basic_command_buffer.h"
+#include "procs.h"
+
+
+//--------------------------------------------------------------------------
+namespace {
+
+// CL_INVALID_COMMAND_BUFFER_KHR if command_buffer is not a valid
+// command-buffer.
+struct FinalizeCommandBufferInvalidCommandBuffer : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    cl_int Run() override
+    {
+        cl_int error = clFinalizeCommandBufferKHR(nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_COMMAND_BUFFER_KHR,
+                               "clFinalizeCommandBufferKHR should return "
+                               "CL_INVALID_COMMAND_BUFFER_KHR",
+                               TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_OPERATION if command_buffer is not in the Recording state.
+struct FinalizeCommandBufferNotRecordingState : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    cl_int Run() override
+    {
+        auto verify_state = [&](const cl_command_buffer_state_khr &expected) {
+            cl_command_buffer_state_khr state = ~cl_command_buffer_state_khr(0);
+
+            cl_int error = clGetCommandBufferInfoKHR(
+                command_buffer, CL_COMMAND_BUFFER_STATE_KHR, sizeof(state),
+                &state, nullptr);
+            test_error_ret(error, "clGetCommandBufferInfoKHR failed",
+                           TEST_FAIL);
+
+            test_assert_error(
+                state == expected,
+                "Unexpected result of CL_COMMAND_BUFFER_STATE_KHR query!");
+
+            return TEST_PASS;
+        };
+
+        cl_int error = RecordCommandBuffer();
+        test_error(error, "RecordCommandBuffer failed");
+        error = verify_state(CL_COMMAND_BUFFER_STATE_EXECUTABLE_KHR);
+        test_error(error, "State is not Executable");
+
+        error = clFinalizeCommandBufferKHR(command_buffer);
+        test_failure_error_ret(error, CL_INVALID_OPERATION,
+                               "clFinalizeCommandBufferKHR should return "
+                               "CL_INVALID_OPERATION",
+                               TEST_FAIL);
+
+        error = EnqueueCommandBuffer();
+        test_error(error, "EnqueueCommandBuffer failed");
+        error = verify_state(CL_COMMAND_BUFFER_STATE_PENDING_KHR);
```

Comment:
I'm not sure the command-buffer will reliably be in the pending state here, as it could have completed and gone back to the executable state. We can block the `clEnqueueCommandBufferKHR` call on a user event instead to guarantee this, for example:
```cpp
cl_event user_event = clCreateUserEvent(context, &error); 

// blocked on user_event 
clEnqueueCommandBufferKHR(0, nullptr, command_buffer, 1,  &user_event, nullptr);

// Verify  pending state
error = verify_state(CL_COMMAND_BUFFER_STATE_PENDING_KHR);

// Test error is CL_INVALID_OPERATION
error = clFinalizeCommandBufferKHR(command_buffer);
  
// Unblock command-buffer execution and wait for completion.
clSetUserEventStatus(user_event, CL_COMPLETE);
clFinish(queue));
```